### DISCUSSION TITLE
How can I read a single line of text that is longer than 1023 characters?

### DISCUSSION THREAD

**Discussion by shogo314** [koka-lang/koka#571] [Q&A] [answered]
The `std/os/readline` function only reads up to 1023 characters. How can I read a longer line of text?

**Comment by TimWhiting** [comment, marked as answer]
You can keep reading (call the function again), but I guess the more specific question you'd have is knowing whether you've read all the stdin / line yet right, since there is no flag letting you know if it read a full line.

For now you can create your own c function that does what you want, and link to it from Koka: Here is all the definitions from `std/os/readline.kk`, `std/os/readline-inline.c`, and the kklib c library `os.c` where this all is implemented.

Extern imports, extern functions etc, are not limited to the std core libraries.  You can use them in your code. (`kklib.h` and all of the imports / types defined in your module are defined prior to your `c` file is spliced in, so you shouldn't have to worry about `c` includes).

```kk
extern import
  c file "readline-inline.c"
  
pub fun readline() : <console,exn> string
  match readline-err() 
    Error(exn)  -> Error(exn.prepend("unable to read from stdin")).untry
    Ok(line)    -> line

extern readline-err() : console error<string>
  c "kk_os_read_line_error"  
  js "sync_readline_err"

```
```c
static kk_std_core_exn__error kk_os_read_line_error( kk_context_t* ctx ) {
  kk_string_t content;
  const int err = kk_os_read_line(&content,ctx);
  if (err != 0) return kk_error_from_errno(err,ctx);
           else return kk_error_ok(kk_string_box(content),ctx);
}

kk_decl_export int kk_os_read_line(kk_string_t* result, kk_context_t* ctx)
{
  char buf[1024];
  if (fgets(buf, 1023, stdin) == NULL) return errno;
  buf[1023] = 0;      // ensure zero termination
  const size_t len = strlen(buf);
  if (len > 0 && buf[len-1] == '\n') {
    buf[len-1] = 0;   // remove possible ending newline character
  }
  *result = kk_string_alloc_from_qutf8(buf, ctx);
  return 0;
}
```
